### PR TITLE
TRef documentation

### PR DESCRIPTION
**Pull request by felipehaack**
First version for the TRef data type documentation.

@mijicd, I'm not sure about the structure (I tried to follow the other examples), would be great to hear your thoughts.

### CODE REVIEW

File: docs/datatypes/tref.md [hunk 1]
```diff
@@ -0,0 +1,223 @@
+---
+id: datatypes_tref
+title:  "TRef"
+---
+
+A `TRef[A]` is part of the `STM` bundle that represents a transactional reference that can perform reads and writes inside a transactional effect.
+
+`TRef` gives the basic machinery for others `SMT` transactional data types.
```
Comment (mijicd) [inline docs/datatypes/tref.md]: Fix typo, `SMT` => `STM`.

File: docs/datatypes/tref.md [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+`TRef[A]` where `A` is an immutable value.
+
+## Create a TRef
+
+Creating a `TRef` without committing its value:
+
+```scala mdoc:silent
+import zio.stm.{STM, TRef}
+
+val tVar: STM[Nothing, TRef[Int]] = TRef.make(10)
```
Comment (mijicd) [inline docs/datatypes/tref.md]: `make` has been renamed to `apply`, hence it's `TRef(10)`.

File: docs/datatypes/tref.md [hunk 3]
```diff
@@ -0,0 +1,223 @@
+---
+id: datatypes_tref
+title:  "TRef"
+---
+
+A `TRef[A]` is part of the `STM` bundle that represents a transactional reference that can perform reads and writes inside a transactional effect.
+
+`TRef` gives the basic machinery for others `SMT` transactional data types.
+
+`TRef[A]` where `A` is an immutable value.
```
Comment (mijicd) [inline docs/datatypes/tref.md]: I would remove this sentence.

File: docs/datatypes/tref.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+## Retrieve the value out of a TRef
+
+By retrieving the value in a single transactional statement: 
+
+```scala mdoc:silent
+import zio.UIO
+import zio.stm.TRef
+
+val tVal: UIO[Int] = (for {
+  tRef <- TRef.make(10)
```
Comment (mijicd) [inline docs/datatypes/tref.md]: See the comment about renaming.

File: docs/datatypes/tref.md [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+```
+
+Or on multiple transactional statements:
+
+```scala mdoc:silent
+import zio.UIO
+import zio.stm.TRef
+
+val tVal: UIO[Int] = for {
+  tRef <- TRef.makeCommit(10)
```
Comment (mijicd) [inline docs/datatypes/tref.md]: I'd consider renaming `makeCommit` method to `committed`. 

@jdegoes wdyt?

File: docs/datatypes/tref.md [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+  value <- tRef.get.commit
+} yield value
+```
+
+## Set a value to a TRef
+
+By setting a value, it will commit a new value without any relation with the actual value at certain point in time.
```
Comment (mijicd) [inline docs/datatypes/tref.md]: I'd rephrase this to something like `Setting the value overwrites the existing content of a reference`

File: docs/datatypes/tref.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+val tVal: UIO[Int] = for {
+  tRef <- TRef.makeCommit(10)
+  _ <- tRef.set(20).commit
+  value <- tRef.get.commit
+} yield value
+```
+
+## Update the value of the TRef
+
+The update function `A => A`, differently from `set`, it cares about the actual value in the `TRef` by performing changes over the value and returning the same type at certain point in time.
```
Comment (mijicd) [inline docs/datatypes/tref.md]: I'd rephrase to `Update modifies the content of a reference by applying function to it`.

File: docs/datatypes/tref.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
+---
+id: datatypes_tref
+title:  "TRef"
+---
+
+A `TRef[A]` is part of the `STM` bundle that represents a transactional reference that can perform reads and writes inside a transactional effect.
```
Comment (jdegoes) [inline docs/datatypes/tref.md]:
```suggestion
A `TRef[A]` is a mutable reference to an immutable value, which can participate in transactions in STM. The mutable reference can be retrieved and set from within transactions, with strong guarantees for atomicity, consistency, and isolation from other transactions.
```

File: docs/datatypes/tref.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+`TRef` gives the basic machinery for others `SMT` transactional data types.
```
Comment (jdegoes) [inline docs/datatypes/tref.md]:
```suggestion
`TRef` provides the low-level machinery to create transactions from modifications of STM memory.
```

File: docs/datatypes/tref.md (hunk 2 above)
Comment (jdegoes) [reply docs/datatypes/tref.md]: Arg, we should change this back (I am sorry I didn't notice that). In Scala, `TRef.apply` should give you a `TRef`, not a `TRef` inside `STM`. So it's a pervasive ZIO custom that `make` gives you something inside an effect.

File: docs/datatypes/tref.md [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+`TRef[A]` where `A` is an immutable value.
+
+## Create a TRef
+
+Creating a `TRef` without committing its value:
```
Comment (jdegoes) [inline docs/datatypes/tref.md]:
```suggestion
Creating a `TRef` inside a transaction:
```

File: docs/datatypes/tref.md [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+
+```scala mdoc:silent
+import zio.stm.{STM, TRef}
+
+val tVar: STM[Nothing, TRef[Int]] = TRef.make(10)
+```
+
+Or creating a `TRef` and immediately committing its value to extract out the value:
```
Comment (jdegoes) [inline docs/datatypes/tref.md]:
```suggestion
Or creating a `TRef` inside a transaction, and immediately committing the transaction, which allows you to store and pass along the reference.
```

File: docs/datatypes/tref.md (hunk 2 above)
Comment (mijicd) [reply docs/datatypes/tref.md]: Gotcha, I'll polish these up asap.

File: docs/datatypes/tref.md (hunk 5 above)
Comment (jdegoes) [reply docs/datatypes/tref.md]: It is `make` followed by `commit`. So `makeCommit`. If we kept with `apply` for `make` then `committed` could work well, or even just `commit`, but I am not a fan of that for reasons listed above.

File: docs/datatypes/tref.md [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+val tVal: UIO[Int] = (for {
+  tRef <- TRef.make(10)
+  _ <- tRef.set(20)
+  nValue <- tRef.get
+} yield nValue).commit
+```
+
+Or on multiple transactional statement:
+
+```scala mdoc:silent
```
Comment (jdegoes) [inline docs/datatypes/tref.md]: This example is not really realistic because each thing is done in isolation from the others, so there are no atomic guarantees across the set and get.

File: docs/datatypes/tref.md (hunk 7 above)
Comment (jdegoes) [inline docs/datatypes/tref.md]:
```suggestion
The update function `A => A` allows computing a new value for the `TRef` using the old value.
```

File: docs/datatypes/tref.md [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+val tVal: UIO[Int] = for {
+  tRef <- TRef.makeCommit(10)
+  _ <- tRef.update(_ + 20).commit
+  value <- tRef.get.commit
+} yield value
+```
+
+## Modify the value of the TRef
+
+The modify function `A => (B, A): B` works similar to `update`, but it gives the possibility to perform computation over the actual value and updating to a new value.
```
Comment (jdegoes) [inline docs/datatypes/tref.md]:
```suggestion
The modify function `A => (B, A): B` works similar to `update`, but allows extracting some information (the `B`) out of the update operation.
```

File: docs/datatypes/tref.md (hunk 4 above)
Comment (felipehaack) [reply docs/datatypes/tref.md]: Sorry, I didn't pull the latest changes. Thanks for noticing. +1

File: docs/datatypes/tref.md (hunk 6 above)
Comment (felipehaack) [reply docs/datatypes/tref.md]: Nice, I prefer your version, clarifies more.

File: docs/datatypes/tref.md (hunk 7 above)
Comment (felipehaack) [reply docs/datatypes/tref.md]: Perfect, let me change to your version.

Comment (felipehaack) [reply docs/datatypes/tref.md]: Thank you for rephrasing and making even more clarified. (same applies for the other comments too)

File: docs/datatypes/tref.md (hunk 12 above)
Comment (felipehaack) [reply docs/datatypes/tref.md]: Sure, you're completely right (also applies for the other examples), for each statement, they're working in isolation. Let me just make sure I'm not wrong, but to have guarantees across the set and get, should I enclosure the statements in a single atomic transaction?
```
for {
  tRef <- TRef(10).commit
  nValue <- STM.atomically {
    for {
      _ <- tRef.set(12)
      nValue <- tRef.get
    } yield nValue
  }
} yield nValue
```
then it gives the guarantees about atomicity of the `set and get`.

File: docs/datatypes/tref.md (hunk 5 above)
Comment (mijicd) [reply docs/datatypes/tref.md]: Reverted the `apply` renaming in #2179. Please take a look.

File: docs/datatypes/tref.md (hunk 4 above)
Comment (felipehaack) [reply docs/datatypes/tref.md]: As you made changes to rolling back to `make`, I will leave as it is.

**Comment by felipehaack** [comment]
@mijicd, thank you for all the comments, they're always welcome.
I pushed the changes just now. Let me know if I'm missing something else.

**Comment by mijicd** [comment]
@felipehaack looks good to me, examples are fixed, changes are incorporated. We can polish it later in case someone notices any issues with the contents.
